File: Ch3/springPendulumAnimation.m
function res = springPendulumAnimation(t,y0,doanim)
% springPendulumAnimation numerically integrates the equations of motion of
% a spring pendulum and optionally animates the results.
%
% res = springPendulumAnimation(t,y0,doanim) performs the integration for
% time array t, with initial conditions y0 ([r(t0),\dot{r}(t0),theta(t0),
% \dot\theta(t0)]) and animates if doanim is true. res contains the results
% of the numerical integration See ode45 for details.

% Copyright (c) 2015 Dmitry Savransky (ds264@cornell.edu)

%spring pendulum Animation
%state y = [r, rd, th,thd]
g = 9.81;   %m/s^2  acceleration due to graivty
x0 = 0.1;   %m      spring rest length
m = 0.1;      %kg     pendulum mass
k = 2;      %N/m    spring constant

if ~exist('t','var') || isempty(t)
    t = 0:1/40:60; %seconds
end
if ~exist('y0','var') || isempty(y0)
    y0 = [0.5,0,pi/3,0];  %initial conditions
end
if ~exist('doanim','var') || isempty(doanim)
    doanim = true;
end

%spring pendulume ODE function
    function dy = springPendulumODE(t,y)
        %y = [r, rd, th,thd]
        r   = y(1);
        rd  = y(2);
        th  = y(3);
        thd = y(4);
        
        dy = [rd;
            r*thd^2 + g*cos(th) - k/m*(r-x0);
            thd;
            -2*rd*thd/r - g*sin(th)/r];
    end

[~,res] = ode45(@springPendulumODE,t,y0,odeset('RelTol',1e-12,'AbsTol',1e-14));

if doanim
    r = res(:,1);
    theta = res(:,3);
    x = r.*sin(theta);
    y = r.*cos(theta);
    
    f = figure(1);
    clf
    
    set(f,'Position',[1,1,1200,700])
    s1 = subplot(2,2,1);
    s2 = subplot(2,2,3);
    s3 = subplot(2,2,[2,4]);
    
    %set up animation
    axes(s3)
    a = 0:pi/100:2*pi;
    circrad = max(r)/20;
    xcirc = circrad*sin(a);
    ycirc = circrad*cos(a);
    mp = fill(xcirc+x(1),ycirc+y(1),'b');
    hold on
    tmp = drawSpring([0,0],[x(1),y(1)],10,0.05);
    sprng = plot(tmp(1,:),tmp(2,:));
    hold off
    axis equal
    axlim = [min(x-circrad),max(x+circrad),...
        min([0;y-circrad]),max([0;y+circrad])];
    axis(axlim)
    
    set(s3,'FontName','Times','FontSize',18,'YDir','Reverse',...
        'XAxisLocation','Top')
    ylabel('$\leftarrow \mathbf{e}_1 $','Interpreter','Latex')
    xlabel('$\mathbf{e}_2 \rightarrow$','Interpreter','Latex')
    
    %set up plots
    axes(s1)
    rplot = plot(t(1),r(1),'Linewidth',2);
    set(s1,'FontName','Times','FontSize',18)
    xlabel('Time (s)','Interpreter','Latex')
    ylabel('$r$ (m)','Interpreter','Latex')
    axes(s2)
    thplot = plot(t(1),theta(1),'Linewidth',2);
    set(s2,'FontName','Times','FontSize',18)
    xlabel('Time (s)','Interpreter','Latex')
    ylabel('$\theta$ (rad)','Interpreter','Latex')
    
    %animate
    for j = 2:length(t)
        pause(t(j)-t(j-1))
        set(mp,'XData',get(mp,'XData')+(x(j) - x(j-1)),...
            'YData',get(mp,'YData')+(y(j) - y(j-1)));
        tmp = drawSpring([0,0],[x(j),y(j)],10,0.05);
        set(sprng,'XData',tmp(1,:),'YData',tmp(2,:));
        
        set(rplot,'XData',t(1:j),'YData',r(1:j));
        set(thplot,'XData',t(1:j),'YData',theta(1:j));
        if t(j) < 5
            set(s1,'Xlim',[0,5])
            set(s2,'Xlim',[0,5])
        else
            set(s1,'XLimMode','auto')
            set(s2,'XLimMode','auto')
        end
    end
end

end

File: engineeringdynamicsdemos/src/engineeringdynamicsdemos/ch3/spring_pendulum_animation.py
"""Spring pendulum numerical integration and animation."""

from typing import Optional

import matplotlib.pyplot as plt
import numpy as np
import numpy.typing as npt
from matplotlib.animation import FuncAnimation
from scipy.integrate import solve_ivp

from ..shared.draw_spring import draw_spring


def spring_pendulum_trajectory(
    t: Optional[npt.ArrayLike] = None,
    y0: Optional[npt.ArrayLike] = None,
    g: float = 9.81,
    x0: float = 0.1,
    m: float = 0.1,
    k: float = 2.0,
    rtol: float = 1e-12,
    atol: float = 1e-14,
) -> npt.NDArray[np.float64]:
    """Numerically integrate the equations of motion of a spring pendulum.

    Args:
        t (npt.ArrayLike, optional):
            Time values (s). Defaults to ``0:1/40:60``.
        y0 (npt.ArrayLike, optional):
            Initial state ``[r, rdot, theta, thetadot]`` (m, m/s, rad, rad/s).
            Defaults to ``[0.5, 0, pi/3, 0]``.
        g (float):
            Acceleration due to gravity (m/s^2). Defaults to 9.81.
        x0 (float):
            Spring rest length (m). Defaults to 0.1.
        m (float):
            Pendulum mass (kg). Defaults to 0.1.
        k (float):
            Spring constant (N/m). Defaults to 2.0.
        rtol (float):
            Relative integration tolerance. Defaults to 1e-12.
        atol (float):
            Absolute integration tolerance. Defaults to 1e-14.

    Returns:
        npt.NDArray[np.float64]:
            Array of shape ``(len(t), 4)`` whose columns are r, rdot, theta and
            thetadot at each time in ``t``.
    """
    if t is None:
        t = np.arange(0, 60 + 1 / 80, 1 / 40)
    if y0 is None:
        y0 = [0.5, 0.0, np.pi / 3, 0.0]
    t = np.asarray(t, dtype=np.float64)

    def spring_pendulum_ode(_t: float, y: npt.NDArray[np.float64]):
        """Right-hand side of the spring pendulum equations of motion.

        Args:
            _t (float):
                Time (s). Unused.
            y (npt.NDArray[np.float64]):
                State ``[r, rdot, theta, thetadot]``.

        Returns:
            list:
                Time derivative of the state.
        """
        r, rd, th, thd = y
        return [
            rd,
            r * thd**2 + g * np.cos(th) - k / m * (r - x0),
            thd,
            -2 * rd * thd / r - g * np.sin(th) / r,
        ]

    sol = solve_ivp(
        spring_pendulum_ode,
        (t[0], t[-1]),
        np.asarray(y0, dtype=np.float64),
        t_eval=t,
        method="DOP853",
        rtol=rtol,
        atol=atol,
    )
    return sol.y.T


def spring_pendulum_animation(
    t: Optional[npt.ArrayLike] = None,
    y0: Optional[npt.ArrayLike] = None,
) -> FuncAnimation:
    """Integrate and animate the motion of a spring pendulum.

    A schematic of the pendulum (spring drawn with
    :func:`engineeringdynamicsdemos.shared.draw_spring`; e1 points down and e2
    points right) alongside live traces of r and theta.

    Args:
        t (npt.ArrayLike, optional):
            Time values (s) of the frames. Defaults to ``0:1/40:60``.
        y0 (npt.ArrayLike, optional):
            Initial state ``[r, rdot, theta, thetadot]``. Defaults to
            ``[0.5, 0, pi/3, 0]``.

    Returns:
        matplotlib.animation.FuncAnimation:
            The animation driving the figure. Callers must keep a reference
            to it alive for it to play.

    Example:
        >>> import matplotlib.pyplot as plt
        >>> anim = spring_pendulum_animation()
        >>> plt.show()
    """
    if t is None:
        t = np.arange(0, 60 + 1 / 80, 1 / 40)
    t = np.asarray(t, dtype=np.float64)
    res = spring_pendulum_trajectory(t, y0)
    r = res[:, 0]
    theta = res[:, 2]
    x = r * np.sin(theta)
    y = r * np.cos(theta)

    fig = plt.figure(1, figsize=(12, 7))
    fig.clf()
    gs = fig.add_gridspec(2, 2, wspace=0.35, hspace=0.35)
    ax_r = fig.add_subplot(gs[0, 0])
    ax_th = fig.add_subplot(gs[1, 0])
    ax_pend = fig.add_subplot(gs[:, 1])

    circ_ang = np.linspace(0, 2 * np.pi, 201)
    circrad = r.max() / 20
    xcirc = circrad * np.sin(circ_ang)
    ycirc = circrad * np.cos(circ_ang)
    (bob,) = ax_pend.fill(xcirc + x[0], ycirc + y[0], "b")
    spr0 = draw_spring((0.0, 0.0), (x[0], y[0]), 10, 0.05)
    (spring_line,) = ax_pend.plot(spr0[0, :], spr0[1, :])
    ax_pend.set_aspect("equal")
    ax_pend.set_xlim(np.min(x - circrad), np.max(x + circrad))
    ax_pend.set_ylim(
        np.max(np.append(y + circrad, 0.0)), np.min(np.append(y - circrad, 0.0))
    )
    ax_pend.xaxis.tick_top()
    ax_pend.xaxis.set_label_position("top")
    ax_pend.tick_params(labelsize=18)
    ax_pend.set_ylabel(r"$\leftarrow \mathbf{\hat{e}}_1$", fontsize=18)
    ax_pend.set_xlabel(r"$\mathbf{\hat{e}}_2 \rightarrow$", fontsize=18)

    (r_line,) = ax_r.plot(t[:1], r[:1], linewidth=2)
    ax_r.tick_params(labelsize=18)
    ax_r.set_xlabel("Time (s)", fontsize=18)
    ax_r.set_ylabel(r"$r$ (m)", fontsize=18)
    margin = 0.1 * np.ptp(r)
    ax_r.set_ylim(r.min() - margin, r.max() + margin)

    (th_line,) = ax_th.plot(t[:1], theta[:1], linewidth=2)
    ax_th.tick_params(labelsize=18)
    ax_th.set_xlabel("Time (s)", fontsize=18)
    ax_th.set_ylabel(r"$\theta$ (rad)", fontsize=18)
    margin = 0.1 * np.ptp(theta)
    ax_th.set_ylim(theta.min() - margin, theta.max() + margin)

    def update(frame: int):
        """Advance the animation to time step ``frame``.

        Args:
            frame (int):
                Index into ``t`` of the animation frame to draw.

        Returns:
            tuple:
                bob (matplotlib.patches.Polygon):
                    The updated pendulum bob.
                spring_line (matplotlib.lines.Line2D):
                    The updated spring.
                r_line (matplotlib.lines.Line2D):
                    The updated r trace.
                th_line (matplotlib.lines.Line2D):
                    The updated theta trace.
        """
        bob.set_xy(np.column_stack((xcirc + x[frame], ycirc + y[frame])))
        spr = draw_spring((0.0, 0.0), (x[frame], y[frame]), 10, 0.05)
        spring_line.set_data(spr[0, :], spr[1, :])
        r_line.set_data(t[: frame + 1], r[: frame + 1])
        th_line.set_data(t[: frame + 1], theta[: frame + 1])
        xmax = 5 if t[frame] < 5 else t[frame]
        ax_r.set_xlim(0, xmax)
        ax_th.set_xlim(0, xmax)
        return bob, spring_line, r_line, th_line

    dt = float(np.mean(np.diff(t)))
    return FuncAnimation(
        fig, update, frames=range(1, len(t)), interval=dt * 1000, blit=False
    )
